middle_end/flambda/un_anf.ml

```diff
@@ -122,32 +122,32 @@ let make_var_info (clam : Clambda.ulambda) : var_info =
   let rec loop ~depth ~environment_vars : Clambda.ulambda -> unit = function
     (* No underscores in the pattern match, to reduce the chance of failing
        to traverse some subexpression. *)
-    | Uvar var -> add_use t var depth
+    | Uvar var -> add_use t var depth; environment_vars
     | Uconst const ->
       (* The only variables that might occur in [const] are those in constant
          closures---and those are all bound by such closures.  It follows that
          [const] cannot contain any variables that are bound in the current
          scope, so we do not need to count them here.  (The function bodies
          of the closures will be traversed when this function is called from
          [Flambda_to_clambda.to_clambda_closed_set_of_closures].) *)
-      ignore_uconstant const
+      ignore_uconstant const; environment_vars
     | Udirect_apply (label, args, dbg) ->
       ignore_function_label label;
-      List.iter (loop ~depth ~environment_vars) args;
-      ignore_debuginfo dbg
+      List.fold_left (loop ~depth ~environment_vars) args;
```

Comment:
I assume you're getting a type error here; you'll need to tweak it a bit. (Note: this applies almost exactly to the other `List.fold_left` occurrences)
`loop` has type `depth:int -> environment_vars:Var.Set.t -> Clambda.ulambda -> Var.Set.t`, and `List.fold_left` has type `('a -> 'b -> 'a) -> 'a -> 'b list -> 'a`.
In this context, `'a` will be `Var.Set.t` (the set of environment vars), `'b` will be `Clambda.ulambda`. So the arguments of `List.fold_left` must be:
- First, a function of type `Var.Set.t -> Clambda.ulambda -> Var.Set.t`. `loop ~depth` would almost fit this purpose, since it has type `environment_vars:Var.Set.t -> Clambda.ulambda -> Var.Set.t`, and I think the compiler would allow you to use it (it knows that function types with labeled arguments are compatible with unlabeled ones), but I would suggest writing your own function instead: `(fun environment_vars arg -> loop ~depth ~environment_vars arg)`
- Then, the initial value of type `Var.Set.t`, in this case the choice of `environment_vars` should be more or less obvious.
- Finally, the list of `Clambda.ulambda` terms, `args` in this case.

You also need to recover and return the result from `List.fold_left`, so you should either switch the `ignore` and `List.fold_left` lines, or bind the result of `List.fold_left` to a variable.

Here's what I would do:
```ocaml
let environment_vars =
  List.fold_left
    (fun environment_vars arg -> loop ~depth ~environment_vars arg)
    environment_vars
    args
in
ignore_debuginfo dbg;
environment_vars
```